keras_preprocessing/image.py

```diff
@@ -1665,6 +1813,18 @@ def __init__(self, directory, image_data_generator,
                  follow_links=False,
                  subset=None,
                  interpolation='nearest'):
+        if dataframe is not None:
```

Comment:
I would prefer if those 2 cases were in different classes. This is not easily maintainable.

Reply:
Maybe we should do a base class with abstract methods that get the classes, samples and filenames.

Reply:
I'm thinking something like a new ImageFileIterator as a Class inherited from Iterator class, which handles all the common stuff except classes, filenames, sample,class_mode and class_indices, and Finally DirectoryIterator and a DataFrameIterator inheriting from ImageFileIterator. Or we can make the Iterator class itself handle all the common stuff, and DirectoryIterator and a DataFrameIterator inheriting from it.

Reply:
Whichever methods you think is the best. I just don't think we should have any :+1: 

```
if dataframe is not None:
   # pandas stuff
elif h5pystuff is not None:
   # H5PY stuff
else:
   # Flow from directory
```

I think it would be better to have one class per Iterator type that handle all everything specific with this Iterator.

Reply:
I have updated my PR, Please have a look.